dhis-2/dhis-web/dhis-web-commons/src/main/java/org/hisp/dhis/servlet/filter/AppOverrideFilter.java

```diff
@@ -0,0 +1,177 @@
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+import lombok.extern.slf4j.Slf4j;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.List;
+import java.util.regex.Pattern;
+import java.util.regex.Matcher;
+
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.core.io.Resource;
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Component;
+import org.springframework.util.StreamUtils;
+import org.springframework.web.context.request.ServletWebRequest;
+
+import org.hisp.dhis.util.DateUtils;
+import org.hisp.dhis.appmanager.App;
+import org.hisp.dhis.appmanager.AppManager;
+import org.hisp.dhis.appmanager.AppStatus;
+
+/**
+ * @author Austin McGee <austin@dhis2.org>
+ */
+ @Slf4j
+ @Component
+public class AppOverrideFilter
+    implements Filter
+{
+    @Autowired
+    private AppManager appManager;
+
+    @Autowired
+    private ObjectMapper jsonMapper;
+
+    public static final String APP_PATH_PATTERN = "^/dhis-web-(" + String.join("|", AppManager.BUNDLED_APPS) + ")/(.*)";
+
+    // -------------------------------------------------------------------------
+    // Filter implementation
+    // -------------------------------------------------------------------------
+
+    @Override
+    public void init( FilterConfig config )
+    {
+    }
+
+    // From AppController.java (some duplication)
+    private void serveInstalledAppResource( App app, String resourcePath, HttpServletRequest request, HttpServletResponse response )
+        throws IOException {
+        // Get page requested
+
+        log.debug( String.format( "Serving app resource: '%s'", resourcePath ) );
+
+        // Handling of 'manifest.webapp'
+        if ( "manifest.webapp".equals( resourcePath ) )
+        {
+            // If request was for manifest.webapp, check for * and replace with host
+            if ( "*".equals( app.getActivities().getDhis().getHref() ) )
```

Comment:
Perhaps some defensive null checks here to see if `activities` is actually defined to avoid NullPointerException, with some logging maybe.

Reply:
Added

Reply:
Also added to the existing code in `AppController.java` which was missing those checks (and copied here)